extensions-core/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentPusher.java

```diff
@@ -99,13 +103,17 @@ public DataSegment push(File inDir, DataSegment segment) throws IOException
       size = CompressionUtils.zip(inDir, out);
     }
 
-    return createDescriptorFile(
+    DataSegment dataSegment = createDescriptorFile(
         segment.withLoadSpec(makeLoadSpec(outFile))
                .withSize(size)
                .withBinaryVersion(SegmentUtils.getVersionFromDir(inDir)),
         outFile.getParent(),
         fs
     );
+
+    Path outDir = new Path(String.format("%s/%s", config.getStorageDirectory(), storageDir));
+    fs.rename(outFile.getParent(), outDir);
```

Comment:
it needs to handle proper cleanup of the temp directory when rename() call returns false (meaning other task won the race) or in case of IOException.

please add UTs for the error cases.